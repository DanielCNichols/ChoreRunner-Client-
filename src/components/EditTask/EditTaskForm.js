import React, { useState } from 'react';
import ApiService from '../../services/api-service';
import { Fieldset, Label, FormElement, Input, Legend } from '../Form/Form';
import s from './EditTaskForm.module.css';

export default function EditTaskForm({
  task: { id, title, points, household_id },
  handleUpdateTask,
  handleToggle,
  editTask,
}) {
  const [inputs, setInputs] = useState({
    id: id,
    title: title,
    points: points,
  });

  const [error, setError] = useState({
    title: null,
    points: null,
    server: null,
  });

  function handleInputChange(ev) {
    ev.persist();
    setInputs(inputs => ({
      ...inputs,
      [ev.target.name]: ev.target.value,
    }));
  }

  function validateInputs() {
    let { title, points } = inputs;
    let errors = {};
    if (!title) {
      errors.title = 'Task title is required';
    }

    if (!points) {
      errors.points = 'Please assign a point value';
    }

    if (points > 100 || points < 1) {
      errors.points = 'Point value must be between 1 and 100';
    }

    if (points % 1 !== 0) {
      errors.points = 'Point value must be a whole number';
    }

    return errors;
  }

  function resetErrors() {
    setError({
      title: null,
      points: null,
      server: null,
    });
  }

  async function handleEditSubmit(ev) {
    try {
      ev.preventDefault();
      resetErrors();

      let errors = validateInputs();

      if (Object.keys(errors).length) {
        return setError(errors);
      }

      let updated = await ApiService.updateTask(household_id, inputs);
      editTask(updated);
      handleToggle();
    } catch (error) {
      setError({ server: error });
    }
  }

  return (
    <form className={s.editTaskForm} onSubmit={handleEditSubmit}>
      <Fieldset>
        <Legend className="videoGameTitles">Edit Task</Legend>
        <FormElement className={s.formElement}>
          <Label htmlFor="title">Title</Label>
          {error.title && (
            <div className={s.error}>
              <span>{error.title}</span>
            </div>
          )}
          <Input
            type="text"
            name="title"
            value={inputs.title}
            onChange={handleInputChange}
          />
        </FormElement>
        <FormElement className={s.formElement}>
          <Label htmlFor="points">Point Value</Label>
          {error.points && (
            <div className={s.error}>
              <span>{error.points}</span>
            </div>
          )}
          <Input
            type="number"
            name="points"
            value={inputs.points}
            onChange={handleInputChange}
          />
        </FormElement>

        {error.server && (
          <div role="alert" className={s.error}>
            <span>{error.server}</span>
          </div>
        )}
        <div className={s.formButtons}>
          <button type="button" onClick={handleToggle}>
            Cancel
          </button>
          <button type="submit">Update Task</button>
        </div>
      </Fieldset>
    </form>
  );
}
