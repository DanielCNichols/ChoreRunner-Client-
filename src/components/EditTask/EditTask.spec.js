import React from 'react';
import ReactDOM from 'react-dom';
import EditTaskForm from './EditTaskForm';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

describe('PlayerStats component testing', () => {
  const task = {
    id: 1,
    title: 'test',
    points: 3,
    household_id: 2,
  };

  const handleToggle = () => {};
  const editTask = () => {};

  it('renders without crashing', () => {
    const div = document.createElement('div');

    ReactDOM.render(
      <MemoryRouter>
        <EditTaskForm
          task={task}
          handleToggle={handleToggle}
          editTask={editTask}
        />
      </MemoryRouter>,
      div
    );

    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders UI as expected', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <EditTaskForm
            task={task}
            handleToggle={handleToggle}
            editTask={editTask}
          />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
