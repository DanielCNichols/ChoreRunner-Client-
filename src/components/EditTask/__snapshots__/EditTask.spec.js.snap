// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerStats component testing renders UI as expected 1`] = `
<form
  className="editTaskForm"
  onSubmit={[Function]}
>
  <fieldset
    className="Fieldset"
  >
    <legend
      className="Legend videoGameTitles"
    >
      Edit Task
    </legend>
    <div
      className="FormElement formElement"
    >
      <label
        className="Label"
        htmlFor="title"
      >
        Title
      </label>
      <input
        className="Input"
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Clean Room"
        type="text"
        value="test"
      />
    </div>
    <div
      className="FormElement formElement"
    >
      <label
        className="Label"
        htmlFor="points"
      >
        Point Value
      </label>
      <input
        className="Input"
        id="points"
        name="points"
        onChange={[Function]}
        placeholder="2"
        type="number"
        value={3}
      />
    </div>
    <div
      className="formButtons"
    >
      <button
        className="arcadeButton"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
      <button
        className="arcadeButton"
        type="submit"
      >
        Update Task
      </button>
    </div>
  </fieldset>
</form>
`;
